### PR TITLE
Working on supporting template with matrix example

### PR DESCRIPTION
**Pull request by ansharlubis**

### CODE REVIEW

File: integration_tests/template_matrix.f90 [hunk 1]
```diff
@@ -70,9 +70,14 @@ elemental function mult_integer(x, y) result(z)
...
-    instantiate matrix_t(integer, plus_integer, mult_integer, m), &
-      only: plus_integer_matrix => plus_matrix
+    integer, parameter :: n = 2
+    instantiate matrix_t(integer, plus_integer, mult_integer, n), &
+      only: integer_matrix => matrix, &
+            integer_plus_matrix => plus_matrix
+    type(integer_matrix) :: m1, m2, m3
+    m1%elements = [[1,0],[0,1]]
+    m2%elements = [[1,2],[2,1]]
+    print *, m1%elements
```
Comment (ansharlubis) [inline integration_tests/template_matrix.f90:80]: Printing this results in a completely random number. Not sure why.

Comment (ansharlubis) [reply integration_tests/template_matrix.f90:80]: Seems like the assignment `m1%elements = [[1,0],[0,1]]` is not allowed yet.

**Review by certik — APPROVED** [review]
I think that this is fine. Let's merge it and we can iterate upon it once merged. Thanks!

File: integration_tests/template_matrix.f90 [hunk 2]
```diff
@@ -69,10 +85,43 @@ elemental function mult_integer(x, y) result(z)
...
+    m2%elements(2,1) = 2
+    m2%elements(2,2) = 1
+
+    call integer_plus_matrix(m1, m2, m3)
+    call integer_mult_matrix(m2, m3, m4)
+
+    print *, m1%elements
+    print *, m2%elements
+    print *, m3%elements
+    print *, m4%elements
```
Comment (certik) [inline integration_tests/template_matrix.f90:123]: I would add here tests like `if (m4%elements(1,1) /= 4) error stop`

**Pull request merged**